netlogo-core/src/main/fileformat/PlotConverter.scala

```diff
@@ -5,45 +5,85 @@ package org.nlogo.fileformat
-    names.groupBy(_.toUpperCase)
-      .toSeq
+  private[fileformat] def generateRenameMappings(groupedNames: Seq[(String, Seq[String])]):  Seq[(String, String)] = {
+    groupedNames
       .map {
-        case (upper, originals) => upper -> originals.distinct
+        case (upper, originals) => upper -> originals
       }
       .flatMap {
         case (upper, originals) if originals.length > 1 =>
           // Put lower-case names before uppercase names
           val sortedOriginals = originals.sorted.reverse
           sortedOriginals.zipWithIndex.map {
             case (original, index) if index == 0 => (original, original)
-            case (original, index) => (original, s"${original}_${index}")
+            case (original, index) => {
+              var new_name = s"${original}_${index}"
+              var new_index = index
+              while(groupedNames.map(_._1).contains(new_name.toUpperCase)){
+                new_index = new_index + 1
+                new_name  = s"${original}_${new_index}"
+              }
+              (original, s"${original}_${new_index}")
+            }
           }
         case _ => Seq.empty[(String, String)]
       }
   }
 
-  private def clarifyProcedureBody(renamePairs: Seq[(String, String)]): String = {
+  private[fileformat] def determinePenSubstitutions(names: Seq[(String, Seq[String])]): Seq[(String, Seq[(String, String)])] = {
+    names.map(plotAndPens =>
+        (plotAndPens._1, generateRenameMappings(plotAndPens._2.groupBy(_.toUpperCase).toSeq))).filterNot(_._2.isEmpty)
+  }
+
+  private[fileformat] def determinePlotSubstitutions(names: Seq[String]): Seq[(String, String)] = {
+    generateRenameMappings(names.groupBy(_.toUpperCase).toSeq)
+  }
+
+  private def clarifyPlotProcedureBody(renamePairs: Seq[(String, String)]): String = {
     val nameMap = renamePairs.map(p => s"""["${p._1}" "${p._2}"]""").mkString("[", " ", "]")
-    s"""|  let name-map ${nameMap}
-        |  let replacement filter [ rename -> first rename = name] name-map
-        |  let reported-name name
-        |  if not empty? replacement [
-        |    set reported-name item 1 replacement
-        |  ]
-        |  report reported-name""".stripMargin
+    s"""    |  let name-map ${nameMap}
+    |  let replacement filter [ rename -> first rename = name ] name-map
+    |  let reported-name name
+    |  if not empty? replacement [
+    |    set reported-name item 1 (item 0 replacement)
+    |  ]
+    |  report reported-name""".stripMargin
+  }
+
+
+  private def clarifyPensProcedureBody(
+    renamePlotPairs: Seq[(String,String)],
+    renamePairs: Seq[(String, Seq[(String, String)])]): String = {
+    val nameMap = renamePlotPairs.map(p => s"""["${p._1}" "${p._2}"]""").mkString("[", " ", "]")
+    val nameReMap = renamePairs.flatMap( renamePair =>
+      renamePair._2.map(p => s"""[ "${renamePair._1}" "${p._1}" "${p._2}" ]""")).mkString("[", " ", "]")
+
+    s"""    |  let plot-name-map ${nameMap}
```

Comment:
It seems like we should be able to check if `renamePlotPairs` is empty - and if it is just use `plot-name` as the `reported-plot`, and otherwise if there were plot renames, just call the plot rename procedure `_clarify-duplicate-plot-name` instead of re-inserting the code for it here.  

If we do keep the code this way, then I think `nameMap` and `nameReMap` might need better names.